ImperatorToCK3/Source/CK3/Province/ProvinceDetails.cpp

```diff
@@ -1,42 +1,30 @@
 #include "ProvinceDetails.h"
 #include "Log.h"
 #include "OSCompatibilityLayer.h"
-#include "ParserHelpers.h"
 #include "CommonRegexes.h"
 
-CK3::ProvinceDetails::ProvinceDetails(const std::string& filePath)
-{
-	registerKeys();
-	if (commonItems::DoesFileExist(filePath))
-	{
-		parseFile(filePath);
-	}
-	else Log(LogLevel::Error) << "Could not open " << filePath << " to load province details.";
-	clearRegisteredKeywords();
-}
 
-void CK3::ProvinceDetails::updateWith(const std::string& filePath)
-{
-	registerKeys();
-	if (commonItems::DoesFileExist(filePath))
-	{
-		parseFile(filePath);
-	}
-	else Log(LogLevel::Error) << "Could not open " << filePath << " to update province details.";
-	clearRegisteredKeywords();
-}
 
-CK3::ProvinceDetails::ProvinceDetails(std::istream& theStream)
-{
-	registerKeys();
-	parseStream(theStream);
-	clearRegisteredKeywords();
-}
+History::Factory CK3::ProvinceDetails::historyFactory = History::Factory({
+	{"culture", "culture", std::nullopt},
+	{"religion", "religion", std::nullopt},
+	{"holding", "holding", "none"}
+});
 
-void CK3::ProvinceDetails::registerKeys()
+CK3::ProvinceDetails::ProvinceDetails(std::istream& theStream)
 {
-	registerSetter("culture", culture);
-	registerSetter("religion", religion);
-	registerSetter("holding", holding);
-	registerMatcher(commonItems::catchallRegexMatch, commonItems::ignoreItem);
+	const auto history = historyFactory.getHistory(theStream);
+	const date date{867, 1, 1};
+	const auto cultureOpt = history->getFieldValue("culture", date);
+	const auto religionOpt = history->getFieldValue("religion", date);
+	const auto holdingOpt = history->getFieldValue("holding", date);
+	if (cultureOpt) {
+		culture = *cultureOpt;
+	}
+	if (religionOpt) {
+		religion = *religionOpt;
+	}
+	if (holdingOpt) {
```

Comment:
How does {"holding", "holding", "none"} factor into std::nullopt you're poking here? Maybe I misunderstood what history class returns.

Reply:
In theory the value for "holding" will never be nullopt as "none" is the default value, but IMO it's better to keep the if for consistency.